File: fsharp/run-length-encoding/RunLengthEncoding.fs
module RunLengthEncoding

open System
open System.Text.RegularExpressions

let encode (input: string) =
    Regex("(.)\\1*").Matches(input)
    |> Seq.map (fun m -> m.Value.[0], m.Value.Length)
    |> Seq.map (fun (letter, count) -> if count = 1 then $"{letter}" else $"{count}{letter}")
    |> String.concat ""

let decode (input: string) =
    Regex("(\\d*)(.)").Matches(input)
    |> Seq.map (fun m -> m.Groups.[2].Value.[0], if m.Groups.[1].Length = 0 then 1 else int m.Groups.[1].Value)
    |> Seq.map String
    |> String.concat ""

File: csharp/run-length-encoding/RunLengthEncoding.cs
using System.Linq;
using System.Text.RegularExpressions;

public static class RunLengthEncoding
{
    public static string Encode(string input) => 
        string.Join("", Regex.Matches(input, @"(.)\1*").Select(Encode));

    private static string Encode(Match match) =>
        match.Length == 1 ? match.Value : $"{match.Length}{match.Value[0]}";

    public static string Decode(string input) => 
        string.Join("", Regex.Matches(input, @"(\d+)?(.)").Select(Decode));

    private static string Decode(Match match) =>
        match.Length == 1 ? match.Value : new string(match.Groups[2].Value[0], int.Parse(match.Groups[1].Value));
}